code/parse/sexp.cpp

```diff
@@ -2543,24 +2552,31 @@ int check_sexp_syntax(int node, int return_type, int recursive, int *bad_node, i
 				break;
 
 			case OPF_AI_GOAL:
+			{
 				if (type2 != OPR_AI_GOAL){
 					return SEXP_CHECK_TYPE_MISMATCH;
 				}
 
-				if (Fred_running) {
-					int ship_num, ship2, w = 0;
+				while (Fred_running) {
+					int ship_num, ship2, wing_num = 0;
 
-					// if it's the "goals" operator, this is part of initial orders, so just assume it's okay
+					// if it's the "goals" operator, this is part of initial orders, so we can't grab the ship from it
 					if (get_operator_const(op_node) == OP_GOALS_ID) {
 						break;
 					}
 
-					ship_num = ship_name_lookup(CTEXT(Sexp_nodes[op_node].rest), 1);	// Goober5000 - include players
+					auto ship_node = Sexp_nodes[op_node].rest;
```

Comment:
The other `ship_node` variable is confined within a separate scope, though.  They are two different instances of the same purpose, like reusing `i` in two different `for` loops.

Checking that `ship_node` >= 0 isn't necessary here because that must have already been true in order for the parser to get to this point.  In any case, it's not done in the similar syntax checking block for subsystems.